### PR TITLE
Add derive, typed lambdas, FQN slash callee, and Elvis-on-ternary Closes #88

### LINKED ISSUE #88: Add `derive` for named pure expressions using typed lambda syntax
## Background

As Sentrie policies grow, authors repeatedly retype the same boolean predicates and small
transformation logic across rules and policies. This causes:

- Copy-paste drift (same logic diverges over time)
- Lower readability (rules become "boolean soup")
- Worse explainability (traces show raw expressions instead of intent)
- More review risk (parameter order mistakes, subtle refactor breakages)
- Missed optimisation opportunities (harder to share and memoize common subgraphs)

We need a first-class way to define and reuse pure expressions without turning the language
into a general-purpose programming language, and without introducing a second calling
convention.

This proposal revises the [original](#45) `derive` design (which used a bespoke `using { ... }`
invocation form) to unify `derive` with the lambda syntax introduced in #77 / #78. The
result is a single callable model for the entire language.

---

## Goals

- Define pure, deterministic named expressions at namespace or policy scope.
- Invoke derived expressions using ordinary call syntax - no second invocation model.
- Optionally type lambda parameters and return values; untyped params accept any value.
- Support namespace-level derives that can be exported and used across namespaces, mirroring
  the shapes model.
- Support policy-level derives for local reuse within the core language.
- Support optional parameters using `?` and defaults via the Elvis operator (`?:`) inside
  bodies.
- Allow single-parameter derives to be passed directly as callbacks to higher-order builtins
  (`filter`, `collect`, `any`, `all`, `first`).
- Preserve determinism, traceability, and clear dependency graphs.
- Provide strong diagnostics: arity mismatch, type mismatch, purity violations, cycles,
  and shadowing.

---

## Non-goals

- Do not allow policy-level derives to be exported.
- Do not introduce a new cross-policy import mechanism (exports are namespace-level symbols
  referenced by fully qualified name, same as shapes).
- Do not change runtime function resolution or TypeScript fall-through behaviour.
- Do not support recursion in derives.
- Do not allow derives to implicitly capture ambient policy execution context (facts, inputs,
  attachments, imported decisions).
- Do not support expression-bodied lambda shorthand in v1 (body must be a block with `yield`).
- **Do not allow TypeScript (module) calls inside any derive - namespace or policy level.**
  If TypeScript logic needs to be reused, it belongs in the rule body or composed via a
  module function called directly from the rule, not abstracted into a derive. See
  [[Purity and Determinism Guardrails](https://claude.ai/chat/49054035-7e7a-47fb-b072-7cac7b0b18b0#purity-and-determinism-guardrails)](#purity-and-determinism-guardrails) for rationale.

---

### Derive definition - namespace or policy scope

```
derive <name> = (<params>) => { <block> }
derive <name> = (<params>): <return_type> => { <block> }
```

Parameters follow the same typed-lambda rules introduced below: a type annotation is
optional. If provided it is validated; if omitted the parameter accepts any value.

**Examples:**

```
// Typed params and return type
derive is_privileged_role = (roles: list<string>): bool => {
  yield "admin" in roles or "ops" in roles
}

// Untyped - accepts any value
derive is_valid_item = (item) => {
  yield item.status == "active"
}

// Optional param with Elvis default
derive threshold_check = (value: number, min?: number): bool => {
  yield value >= (min ?: 0)
}

// Multi-param with mixed typing
derive full_name = (first: string, last: string): string => {
  yield first + " " + last
}
```

### Derive invocation - ordinary call syntax

```
yield is_privileged_role(user.roles)
yield threshold_check(score, 10)
yield full_name(user.first, user.last)
```

No new invocation form. Derives are called exactly like builtins and module functions.

Name resolution order at a call site:
1. Derives in scope (policy-local first, then namespace-level)
2. Built-in functions
3. TypeScript fall-through

This matches the existing function resolution model.

### Fully qualified invocation - cross-namespace

```
yield com/example/iam/is_privileged_role(user.roles)
```

No additional import statement. Fully qualified reference is the mechanism, same as shapes.

### Passing a derive as a callback

Single-parameter derives are already `ValueCallable` values and can be passed directly to
higher-order builtins:

```
// is_valid_item has arity 1 - passes directly
filter(items, is_valid_item)

// get_key has arity 1 - works with distinct
distinct(users, get_key)
```

Multi-parameter derives must be wrapped in an anonymous lambda:

```
// is_above_threshold has arity 2 - wrap it
filter(scores, (s) => { yield is_above_threshold(s.value, 50) })
```

---

## Typed Lambda Parameters

This change generalises the lambda syntax to support optional type annotations on parameters.
This applies to both named derives and anonymous lambdas everywhere in the language.

### Syntax

```
// Untyped (existing - unchanged)
(item) => { yield item.value }

// Typed
(item: User) => { yield item.role == "admin" }

// Mixed
(item: User, idx) => { yield item.role == "admin" }

// Optional typed param
(value: number, min?: number) => { yield value >= (min ?: 0) }

// Optional untyped param
(value, fallback?) => { yield value ?: fallback }
```

### Rules

- Type annotations are optional on every parameter individually.
- If a type annotation is present, the runtime validates the supplied argument against the
  declared type before executing the body. A type mismatch is a runtime error with a useful
  trace showing the expected type, actual value, and parameter name.
- If no type annotation is present, the parameter accepts any value (same semantics as today).
- Optional parameters (`?`) bind to `undefined` when not supplied at the call site.
- Optional parameters must follow all required parameters in the list.
- Return type annotation (`: type`) is optional. If present, the yielded value is validated
  against it.
- These rules apply uniformly to anonymous lambdas passed to builtins and to lambdas used
  as derive bodies.

---

## Export Semantics - Namespace Derives Only

Namespace derives may be exported using a standalone export statement, identical to shape
exports:

```
export derive is_privileged_role
```

Rules:
- Only derives defined at namespace scope may be exported.
- Policy-level derives cannot be exported.
- The export statement must refer to a namespace-level derive in the same compilation unit.
- Exported derives become part of the namespace's public API and must remain pure and
  portable.

---

### Namespace-level derives

- Visible to all policies within the same namespace.
- May be exported.
- Purity enforced strictly:
  - No module calls (modules exist only in policies).
  - No decision imports.
  - No free variables resolving to ambient policy execution context.
- All required data must arrive via explicit parameters.
- Cycles rejected at load/compile time.

### Policy-level derives

- Visible only within that policy.
- Cannot be exported.
- May reference namespace derives (local or exported via fully qualified name).
- May reference other policy-local derives.
- Must not call TypeScript module functions. Derives at any scope are strictly
  core-language. If TypeScript logic needs reuse, call it in the rule body directly or
  compose it via a module function at the rule level - not inside a derive.
- Must not implicitly capture ambient policy execution context. Facts, inputs, attachments,
  and imported decision outputs must be passed explicitly as arguments.

Purity is enforced by static analysis on the lambda body at the `derive` binding site: the
body is walked for free variables; any identifier resolving outside the lambda's own param
scope to ambient context is a compile error.

### Shadowing

- Policy-local derive names may shadow namespace derives of the same name within that policy.
- Fully qualified name always allows unambiguous access to an exported derive.
- Diagnostics should make shadowing visible in explain traces and error messages.

---

## Purity and Determinism Guardrails

Derived expressions are pure DAG nodes. Enforce:

- No side effects.
- No ambient state: no implicit access to time, randomness, IO, environment, or external
  state. If time is needed, pass it explicitly as a parameter.
- No callable return values: a derive must not return a `ValueCallable`. Returning a
  callable would allow ambient execution context to leak through the closure.
- These rules apply to **both** namespace-level and policy-level derives.

### TypeScript calls are prohibited in all derives

TypeScript module functions are opaque to the static analyser and can perform side effects,
network calls, or non-deterministic operations. Allowing TS calls inside a derive would
make it impossible to guarantee that derives are safe to memoize, produce reliable traces,
or be trusted as auditable policy logic.

The rule is simple: **no derive at any scope may call a TypeScript module function.**

If TypeScript logic needs to be reused across rules, the correct pattern is to call it
directly in the rule body, or to compose it at the rule level after the derive has produced
its output:

```
// Correct - TS call stays in the rule body, derive stays pure
rule allow = {
  let enriched = enrich_user(user)          // TS call here
  yield is_privileged_role(enriched.roles)  // derive called with the result
}

// Incorrect - TS call hidden inside a derive
derive is_privileged_role = (user_id: string): bool => {
  let user = fetch_user(user_id)  // TS call - rejected at compile time
  yield "admin" in user.roles
}
```

The purity checker must resolve call sites during static analysis (not just walk for free
variables) so that TS fallthrough calls can be detected and rejected, not just ambient
context captures.

---

## Evaluation and Explainability

Derive invocation is evaluated as part of the expression DAG.

Traces use param names even though invocation is positional:

```
is_privileged_role(roles=["admin", "dev"]) → true
threshold_check(value=42, min=50) → false
```

The param names come from the lambda AST. No special trace format is needed - the existing
callable trace machinery emits them.

Shadowing resolution is noted where a policy-local derive overrides a namespace derive.

---

## Backwards Compatibility

- Runtime function calls remain unchanged. Resolution order is: derives → builtins →
  TypeScript fall-through.
- The `using` keyword is not introduced. No keyword reservation needed beyond `derive`,
  which was already anticipated in the v0.2 milestone.
- Anonymous lambdas with untyped params behave identically to today.
- Type annotations on lambda params are purely additive - no existing code changes.

---

## Parser and Precedence

`derive` is a top-level statement in both namespace and policy bodies.

```
DeriveStatement = "derive" IDENT "=" LambdaExpression
```

`LambdaExpression` is extended to support typed params:

```
LambdaExpression = "(" TypedParamList? ")" ReturnTypeAnnotation? "=>" BlockExpression
TypedParamList   = TypedParam ("," TypedParam)*
TypedParam       = IDENT "?"? (":" TypeRef)?
ReturnTypeAnnotation = ":" TypeRef
```

The existing grouped-expression vs lambda disambiguation in `parser/block.go` and
`parser/lambda.go` extends naturally: `(x: T)` with a colon after the identifier is
unambiguously a typed lambda param, not a grouped expression.

---

### Extend `LambdaExpression`

```go
type LambdaExpression struct {
    *baseNode
    Params     []string   // parameter names (unchanged)
    ParamTypes []TypeRef  // parallel slice; nil entry = untyped
    ParamOpts  []bool     // parallel slice; true = optional (?)
    ReturnType TypeRef    // nil = untyped return
    Body       *BlockExpression
}
```

### Add `DeriveStatement`

```go
type DeriveStatement struct {
    *baseNode
    Name  string
    Value *LambdaExpression
}
```

### Add `ExportDeriveStatement`

```go
type ExportDeriveStatement struct {
    *baseNode
    Name string
}
```

---

### Purity check at derive binding

When a `DeriveStatement` is evaluated (at load/compile time), walk the lambda body and
validate that no free variable resolves to ambient execution context. Emit a targeted
error if it does.

### Typed param validation at call time

When a callable backed by a typed lambda is invoked, validate each typed param against
its declared `TypeRef` before binding. Type mismatch produces a runtime error with:

- Parameter name
- Expected type
- Actual value and kind

### Optional param handling

Untyped optional: bind to `box.Undefined()` if not supplied.
Typed optional: if supplied, validate type; if not supplied, bind to `box.Undefined()`.

### Callable value for derives

Derives produce the same `ValueCallable` as anonymous lambdas. Arity is the number of
required params (optional params do not count for arity matching when passed as callbacks
to higher-order builtins - optional params beyond arity are bound to `undefined`).

### Memoisation (optional)

Derives with only scalar or shape-typed params are candidates for memoisation keyed on a
canonicalised argument map. This is optional in v1.

---

### 1. Lexer / tokens

- `derive` is tokenised as a keyword.
- Typed param syntax (`IDENT : TypeRef`) is lexed correctly in lambda param lists.
- `?` in param position is lexed correctly as the optional marker.

### 2. Parser

- `derive IDENT = LambdaExpression` parses as a `DeriveStatement` at namespace and policy
  scope.
- `export derive IDENT` parses as an `ExportDeriveStatement` at namespace scope.
- Typed params parse correctly in both anonymous lambdas and derive bodies.
- Return type annotation parses correctly.
- Grouped-expression vs lambda disambiguation continues to work with typed params.
- Optional untyped params (`IDENT?`) parse correctly.
- Optional typed params (`IDENT?: TypeRef`) parse correctly.

### 3. AST and Validation

- `DeriveStatement`, `ExportDeriveStatement` nodes exist with correct fields.
- `LambdaExpression` carries `ParamTypes`, `ParamOpts`, and `ReturnType`.
- Validation:
  - Optional params must follow required params.
  - No duplicate param names.
  - No recursion / cycles across derives.
  - Namespace derives: purity check (no free variables resolving to ambient context, no
    callable return values, no TypeScript call sites).
  - Policy derives: no implicit ambient capture; no TypeScript call sites.
  - Export: only namespace-level derives may be exported; policy derives rejected.

### 4. Runtime / Evaluator

- Derives evaluate to `ValueCallable`.
- Typed params are validated at call time with useful error messages.
- Optional params bind to `undefined` when not supplied.
- Single-param derives pass directly as callbacks to `filter`, `collect`, `any`, `all`,
  `first`, `distinct`.
- Purity check runs at derive binding time; ambient-capture errors are reported clearly.
- Existing anonymous lambda behaviour is unchanged.
- Existing untyped call semantics for builtins and module functions are unchanged.

### 5. Exports

- `export derive <name>` makes the derive resolvable via fully qualified name from other
  namespaces.
- Policy-local derives are rejected at export with a helpful error.

### 6. Tooling and Diagnostics

- Arity mismatch at a derive call site: expected N args, got M.
- Type mismatch on a typed param: clear message with param name, expected type, actual
  value.
- Purity violation: clear message identifying the free variable or TypeScript call site
  that violates the derive's purity contract.
- Cycle detection with a useful cycle path.
- Shadowing noted in traces and error messages.
- "Did you mean?" suggestions for near-miss derive names.

### 7. Documentation

- Reference docs for derive definition, typed params, return types, optional params,
  exports, and fully qualified invocation.
- Explicit documentation of the relationship to anonymous lambdas: derives are named,
  pure, scoped lambda bindings - not a separate language concept.
- Explicit documentation of the TypeScript prohibition: derives may not call TS module
  functions at any scope. Document the correct pattern (call TS in the rule body, pass
  the result into a derive).
- Examples covering:
  - Namespace derive with typed params (exported and local usage)
  - Policy-level derive for local reuse
  - Fully qualified derive invocation
  - Optional params with Elvis defaults
  - Single-param derive passed as a callback to a builtin
  - Multi-param derive wrapped in an anonymous lambda for use as a callback
  - Correct pattern for composing TS calls with derives

---

### Policy using a namespace derive

```
namespace iam

derive is_privileged_role = (roles: list<string>): bool => {
  yield "admin" in roles or "ops" in roles
}

export derive is_privileged_role

policy access_check {
  fact user: User

  rule allow = {
    yield is_privileged_role(user.roles)
  }

  export decision of allow
}
```

### Cross-namespace usage

```
namespace billing

policy invoice_access {
  fact user: User

  rule allow = {
    yield com/example/iam/is_privileged_role(user.roles)
  }

  export decision of allow
}
```

### Policy-level derive with optional param

```
policy scoring {
  fact event: Event

  derive adjusted_score = (raw: number, penalty?: number): number => {
    yield raw - (penalty ?: 0)
  }

  rule high_score = {
    yield adjusted_score(event.raw_score, event.penalty) >= 80
  }

  export decision of high_score
}
```

### Single-param derive as a direct callback

```
namespace products

derive is_available = (item: Item): bool => {
  yield item.stock > 0 and item.status == "active"
}

policy catalog {
  fact inventory: list<Item>

  rule has_stock = {
    yield any(inventory, is_available)
  }

  export decision of has_stock
}
```

### Multi-param derive wrapped in a lambda for collection use

```
derive above_threshold = (value: number, min: number): bool => {
  yield value >= min
}

rule qualified = {
  yield filter(scores, (s) => { yield above_threshold(s.value, 50) })
}
```

### Correct pattern for composing TypeScript calls with derives

TypeScript calls must stay in the rule body. The derive receives the result as an explicit
argument - it never calls TS itself.

```
// user_utils is a TypeScript module
use user_utils

derive is_privileged_role = (roles: list<string>): bool => {
  yield "admin" in roles or "ops" in roles
}

policy access_check {
  fact user_id: string

  rule allow = {
    // TS call happens here, in the rule body
    let user = user_utils.fetch_user(user_id)
    // pure derive receives the result as data
    yield is_privileged_role(user.roles)
  }

  export decision of allow
}
```

This pattern keeps derives fully auditable and traceable while leaving TypeScript as the
escape hatch for IO and enrichment - where it belongs.

### PR DESCRIPTION
**Pull request by binaek**
## Summary
- Adds first-class derive parsing, indexing, validation, and runtime invocation support.
- Hardens derive evaluation to enforce purity, export boundaries, and deterministic call behavior.
- Aligns and expands tests/CI to reduce race-test flakiness and keep branch coverage stable.

## What this PR does
- Introduces `derive` language support end-to-end (tokens, AST, parser, index, runtime).
- Enforces derive purity and derive-safe callable execution paths at index/runtime boundaries.
- Tightens CI reliability and modernizes test structures/context usage in touched packages.

### Language and parser
- Adds derive keyword/token support and derive AST nodes.
- Adds typed lambda and slash-callee parsing required by derive call patterns.
- Rejects typed lambdas with duplicate parameter names.
- Extends grammar docs and parser tests for derive, typed lambda, and Elvis ternary paths.

### Indexing and validation
- Indexes derives with ordered `AddProgram` behavior and dependency/cycle checks.
- Adds derive expression walker and purity validator to restrict disallowed identifiers/calls.
- Allows passing single-parameter derives as callbacks to higher-order pure builtins (e.g. `filter(items, predDerive)`).
- Enforces namespace export checks for derive access across namespace boundaries.

### Runtime evaluator and execution
- Resolves and invokes derives via identifier and slash-FQN call paths.
- Wraps derive identifiers as callables for higher-order builtin callback positions.
- Propagates derive execution context through attached child contexts.
- Restricts derive execution from invoking non-derive-safe builtins and module calls.
- Centralizes derive-safe builtin registry and adds coverage around derive invocation edge cases.
- Improves derive argument type-mismatch errors to include parameter names.

### Tests and suite conventions
- Expands derive-centric tests across parser, index, and runtime.
- Aligns runtime/index tests toward package suite conventions and `t.Context()`/`s.T().Context()` usage.
- Refactors expensive derive runtime test setup to reduce repeated parse/index/validate/exec work.
- Updates `index/derive_test.go` to use `IndexTestSuite` methods instead of standalone test functions.

### CI and coverage
- Adds and tunes race/test timeouts and job settings for more stable CI runs.
- Simplifies coverage workflow guardrails and keeps patch coverage informational where needed.
- Adds focused coverage tests for derive hardening and evaluator paths.

### Cursor rules and plans
- Adds/updates Cursor rules for evalDerive resolution and Go test conventions.
- Removes stale derive-related Cursor plan markdown from this branch.

## Review Notes
- Focus review on purity enforcement boundaries in `index/derive_purity.go` and runtime dispatch in `runtime/eval_call.go`.
- Verify derive export checks for both slash-FQN and identifier-based call paths.
- Confirm context propagation (`evalDerive`) is preserved only where intended in execution contexts.
- Recheck test-suite refactors for semantics parity (structure changed, expected behavior unchanged).

## Testing Notes
- `go test ./index`
- `go test -race -p=1 ./runtime`
- `go test -race -p=1 ./...`

## Dependency changes
- No new external dependencies added.

### CODE REVIEW

**Review by binaek — COMMENTED** [review]
## Code review (code-reviewer)

Solid feature slice: derive indexing, cycle detection, detached invoke path, Elvis on ternary, and broad tests. **Two correctness gaps should be fixed before merge:**

### Critical

1. **`evalDerive` is lost in `evalBlock`** (`runtime/eval_block.go:32`) — `invokeDerive` sets `evalDerive` on a detached context, but `evalBlock` immediately wraps with `AttachedChildContext()` which does **not** inherit `evalDerive` (`runtime/exec_ctx.go:112-122`). Derive bodies therefore run with `evalDerive == nil`, disabling define-site resolution, builtin whitelist, export checks, and rule-dispatch blocking.

2. **Cross-namespace export only when `evalDerive != nil`** (`runtime/eval_call.go:156-170`) — export checks in `lookupDeriveBySlashFQ` only run in the `ec.evalDerive != nil` branch. Combined with (1), rule-level slash calls can hit unexported cross-namespace derives.

### Warnings

- **`runtime/eval_ident.go:70-79`** — rule dispatch blocked only when `evalDerive != nil`; derive bodies can accidentally `execRule` when referencing a rule name.
- **`runtime/eval_derive_test.go:219`** — `TestGetTargetSlashCrossNamespaceDeriveRequiresExport` tests `getTarget` with manually set `evalDerive`, not `invokeDerive` → `evalBlock`; add `ExecRule` integration test for cross-namespace export failure.
- **`index/index.go:101`** — `cloneDeriveMap(ns.Derives)` at policy index time: namespace `derive` after `policy` in same file → define-site vs runtime mismatch.
- **`index/derive_cycle.go:47`** — `deriveCallees` omits some expression kinds; cycles through those forms may pass validation.

### Suggestions

- Inherit `evalDerive` in `AttachedChildContext` rather than duplicating checks in multiple call sites.
- Extract shared cross-namespace export helper for rule and derive paths.
- Consider dropping `.cursor/plans/derive_feature_88_e3809fda.plan.md` from merge unless repo policy keeps Cursor plans on main.

### Test recommendations

- `ExecRule`: cross-namespace slash derive must fail with `NotExportedError` after `evalDerive` propagation fix.
- Derive body calls disallowed builtin; derive body references rule name (must not execute rule).
- Index: namespace derive declared after policy — lock expected behavior with test.

File: runtime/derive_invoke.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	if err := child.PushRefStack(d.FQN.String()); err != nil {
+		return box.Undefined(), err
+	}
+	defer child.PopRefStack()
+
+	for i, name := range lam.Params {
+		child.SetLocal(name, args[i], true)
+	}
+
+	val, _, err := evalBlock(ctx, child, exec, callerPolicy, lam.Body)
```
Comment (binaek) [inline runtime/derive_invoke.go:75]: **Critical:** `invokeDerive` sets `evalDerive` on the detached context, then calls `evalBlock`. But `evalBlock` (`runtime/eval_block.go:32`, unchanged in this PR) always wraps with `AttachedChildContext()`, which does **not** inherit `evalDerive` — so the EC actually evaluating the lambda body has `evalDerive == nil`.

**Fix:** inherit `evalDerive` in `AttachedChildContext`, or avoid `AttachedChildContext` when already in derive mode.

File: runtime/eval_call.go [hunk 2]
```diff
@@ -127,17 +127,96 @@ func calculateHashKey(node *ast.CallExpression, args []box.Value) string {
...
+	parts := strings.Split(fqn, ast.FQNSeparator)
+	if len(parts) < 3 {
+		return nil, nil
+	}
+	if ec.evalDerive != nil {
+		if d := ec.evalDerive.DefineFQN[fqn]; d != nil {
+			return d, nil
+		}
+		if d, ok := exec.index.DerivesByFQN[fqn]; ok {
+			if d.Namespace.FQN.String() != ec.evalDerive.Namespace.FQN.String() {
```
Comment (binaek) [inline runtime/eval_call.go]: **Critical:** `VerifyDeriveExported` only runs when `ec.evalDerive != nil` (lines 156-167). With `evalDerive` lost in derive bodies (see `derive_invoke.go`), and with `ec.evalDerive == nil` on rule paths, slash FQN calls can invoke unexported cross-namespace derives via `ResolveDerive` (line 170).

**Fix:** enforce export whenever `d.Namespace.FQN != callerNamespace`.

File: runtime/eval_ident.go [hunk 3]
```diff
@@ -67,14 +67,16 @@ func evalIdent(ctx context.Context, ec *ExecutionContext, exec *executorImpl, p
 		return val, n.SetResult(val), nil
 	}
 
-	if r, found := p.Rules[i.Value]; found {
-		decision, _, node, err := exec.execRule(ctx, ec, p.Namespace.FQN.String(), p.Name, r.Name)
-		n.Attach(node)
-		if err != nil {
-			return box.Undefined(), n.SetErr(err), err
+	if ec.evalDerive == nil {
```
Comment (binaek) [inline runtime/eval_ident.go:79]: **Warning:** Rule dispatch blocked only when `ec.evalDerive != nil`. Because `evalDerive` is lost in derive bodies, a derive referencing a rule name can accidentally run `execRule`.

File: runtime/eval_derive_test.go [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+	exec, err := NewExecutor(idx)
+	s.Require().NoError(err)
+
+	out, err := exec.ExecRule(ctx, "com/ex", "pol", "gate", nil)
+	s.Require().NoError(err)
+	s.Equal(trinary.True, out.Decision.State)
+}
+
+func (s *RuntimeTestSuite) TestGetTargetSlashCrossNamespaceDeriveRequiresExport() {
```
Comment (binaek) [inline runtime/eval_derive_test.go]: **Warning:** Tests `getTarget` with manually set `evalDerive`, not `invokeDerive` → `evalBlock`. Add `ExecRule` integration test: cross-namespace slash derive must fail with `NotExportedError`.

File: index/index.go [hunk 5]
```diff
@@ -79,31 +83,50 @@ func (idx *Index) AddProgram(ctx context.Context, astProgram *ast.Program) error
...
+			if err := ns.addShape(shape); err != nil {
+				return err
+			}
 
-	for _, policy := range program.Policies {
-		p, err := createPolicy(ns, policy, astProgram)
-		if err != nil {
-			return err
-		}
+		case *ast.PolicyStatement:
```
Comment (binaek) [inline index/index.go:101]: **Warning:** `cloneDeriveMap(ns.Derives)` snapshots namespace derives at policy index time. A namespace `derive` declared *after* a `policy` in the same file is omitted from `DefineShort`/`DefineFQN` but may still be reachable at runtime — define-site vs runtime mismatch.

File: index/derive_cycle.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+		ps := make([]string, 0, len(paths))
+		for _, n := range paths {
+			ps = append(ps, n.FQN.String())
+		}
+		return fmt.Errorf("cyclic derive dependency: %s: %w", strings.Join(ps, " -> "), xerr.ErrIndex)
+	}
+	return nil
+}
+
+func deriveCallees(idx *Index, d *Derive) []*Derive {
```
Comment (binaek) [inline index/derive_cycle.go:49]: **Warning:** `deriveCallees` walker omits some expression kinds (e.g. `IsDefinedExpression`, pipeline holes). Cycles through those forms may pass validation.

File: .cursor/plans/derive_feature_88_e3809fda.plan.md [hunk 7]
```diff
@@ -0,0 +1,275 @@
+---
```
Comment (binaek) [inline .cursor/plans/derive_feature_88_e3809fda.plan.md]: **Suggestion:** Drop this Cursor delivery plan from the merge unless the repo explicitly keeps `.cursor/plans/` on main.
